src/main/io/rcdevice.c

```diff
@@ -0,0 +1,1013 @@
+                data[dataPos++] = c;
+
+                settingType = serialRead(device->serialPort);
+                crc = crc8_dvb_s2(crc, settingType);
+                data[dataPos++] = settingType;
+
+                switch (settingType) {
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_UINT8:
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_INT8:
+                    expectedPacketLen = 7; // [value]+[min value]+[max value]+[step size]+[crc], all of them(expect crc) are type of UINT8 or INT8
+                    break;
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_UINT16:
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_INT16:
+                    expectedPacketLen = 11; // [value]+[min value]+[max value]+[step size]+[crc], all of them(expect crc) are type of UINT16 or INT16
+                    break;
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_FLOAT:
+                    expectedPacketLen = 9; // [value:UINT8]+[min value:UINT8]+[max value:UINT8]+[decimal point:UINT16]+[step size:UINT8]+[crc:UINT8]
+                    break;
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_TEXT_SELECTION:
+                    expectedPacketLen = 63; // [value:UINT8]+[text_selections]+[crc:UINT8], text_selections is a null-terminated string
+                    break;
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_STRING:
+                    expectedPacketLen = 63; // [value]+[max string size:UINT8]+[crc:UINT8], value is a null-terminated string
+                    break;
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_FOLDER:
+                case RCDEVICE_PROTOCOL_SETTINGTYPE_INFO:
+                    expectedPacketLen = 3; // expect [remain chunk count]+[setting type], there is only crc field in it
+                    break;
+                default:
+                    return false; // unexpected setting type
+                }
+
+                settingTypeReceived = 1;
+            }
+        } else {
+            if (serialRxBytesWaiting(device->serialPort) > 0) {
+                uint8_t c = serialRead(device->serialPort);
+                crc = crc8_dvb_s2(crc, c);
+                data[dataPos++] = c;
+
+                if (isFoundANullTeminatedChar) {
+                    expecteRemainingDataLen--;
+                    if (expecteRemainingDataLen == 0) {
+                        break;;
+                    }
+                }
+
+                if (settingType == RCDEVICE_PROTOCOL_SETTINGTYPE_TEXT_SELECTION) {
+                    if (!receiveCurrentValueForTextSelectionType) {
+                        receiveCurrentValueForTextSelectionType = true;
+                    } else {
+                        if (c == 0) {
+                            // found the null teminated char, and get next char(crc field), then break the loop
+                            isFoundANullTeminatedChar = true; 
+                            expecteRemainingDataLen = 1;
+                        }
+                    }
```

Comment:
This `if` statement is a double negative, eg `if (!condition) A; else B;` this should be replaced with `if (condition) B; else A;` eg
````
if (receiveCurrentValueForTextSelectionType) {
    if (c == 0) {
        // found the null teminated char, and get next char(crc field), then break the loop
        isFoundANullTeminatedChar = true; 
        expecteRemainingDataLen = 1;
    }
} else {
    receiveCurrentValueForTextSelectionType = true;
}
```